fe/fe-core/src/main/java/com/starrocks/lake/StarMgrMetaSyncer.java

```diff
@@ -208,64 +255,70 @@ private void deleteUnusedShardAndShardGroup() {
-        LOG.debug("diff.size is {}, diff: {}", diffGroupInfoMap.size(), diffGroupInfoMap.keySet());
-
-        // 1.4.collect redundant shard groups and delete
-        List<Long> emptyShardGroup = new ArrayList<>();
-        for (Map.Entry<Long, ShardGroupInfo> entry : diffGroupInfoMap.entrySet()) {
-            long shardGroupId = entry.getKey();
-            ShardGroupInfo shardGroupInfo = entry.getValue();
-
-            if (!isSafeToDelete(shardGroupId, shardGroupInfo)) {
-                continue;
+        Map<Long, ShardGroupInfo> diffGroupInfoMap = new HashMap<>();
+        long nextShardGroupId = 0;
+        do {
+            StarOSAgent.ListShardGroupResult result;
+            try {
+                result = starOSAgent.listShardGroup(nextShardGroupId);
+                nextShardGroupId = result.nextShardGroupId();
+            } catch (DdlException exception) {
+                LOG.info("Fail to list shardgroup from starmgr: {}. Abort the clean up", exception.getMessage());
+                break;
             }
 
-            long createTimeTs = Long.parseLong(entry.getValue().getPropertiesOrDefault("createTime", "0"));
-            if (createTimeTs == 0) {
-                LOG.debug("Can't parse createTime from shardGroup:{} properties, ignore it for now.", shardGroupId);
-                continue;
-            }
+            result.shardGroupInfos().stream()
+                    .filter(x -> x.getGroupId() != 0)
+                    .filter(x -> !groupIdFe.contains(x.getGroupId()))
+                    .forEach(x -> diffGroupInfoMap.put(x.getGroupId(), x));
+
+            if (nextShardGroupId == 0 || diffGroupInfoMap.size() >= SHARD_GROUP_MAX_CLEAN_SIZE) {
+                List<Long> emptyShardGroup = new ArrayList<>();
+                for (Map.Entry<Long, ShardGroupInfo> entry : diffGroupInfoMap.entrySet()) {
+                    long shardGroupId = entry.getKey();
+                    ShardGroupInfo shardGroupInfo = entry.getValue();
+                    if (!isSafeToDelete(shardGroupId, shardGroupInfo)) {
+                        continue;
+                    }
 
-            ComputeResource computeResourceLocal = computeResource;
-            try {
-                long tableId = Long.parseLong(shardGroupInfo.getLabels().get("tableId"));
-                computeResourceLocal = GlobalStateMgr.getCurrentState().getWarehouseMgr().getBackgroundComputeResource(tableId);
-            } catch (Exception e) {
-                LOG.debug("can not get background compute resource, {}", e);
-                // continue, default compute resource is already set
-            }
+                    long createTimeTs = Long.parseLong(shardGroupInfo.getPropertiesOrDefault("createTime", "0"));
+                    if (createTimeTs == 0) {
+                        LOG.debug("Can't parse createTime from shardGroup:{} properties, ignore it for now.", shardGroupId);
+                        continue;
```

Comment:
if here continues, should remove it from diffGroupInfoMap? if not, might be a dead loop

Reply:
2 continue